### PR TITLE
Adding client name to clients for user agent...

### PR DESCRIPTION
**Pull request by spfink**
Renaming SERVICE_NAME to SIGNING_NAME

### CODE REVIEW

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java [hunk 1]
```diff
@@ -31,20 +32,38 @@
 public class ApplyUserAgentStage implements MutableRequestToRequestPipeline {
 
     private final LegacyClientConfiguration config;
+    private final String clientName;
+    private final String clientType;
 
-    public ApplyUserAgentStage(HttpClientDependencies dependencies) {
+    public ApplyUserAgentStage(HttpClientDependencies dependencies,
+                               String clientName,
+                               String clientType) {
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java]: I would forgo the type since that should be obvious from the client name. You can also get the client name from the dependencies so you don't need to do a custom supplier.

File: http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpClient.java [hunk 2]
```diff
@@ -32,4 +32,5 @@
      */
     AbortableCallable<SdkHttpFullResponse> prepareRequest(SdkHttpFullRequest request, SdkRequestContext requestContext);
 
+    String clientName();
```
Comment (shorea) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpClient.java]: We should clearly document this since it has the potential to skew our metrics if people misuse it. Also are there any characters that could screw up parsing of the user agent? Would we need to sanitize.

File: services/rds/src/test/java/software/amazon/awssdk/services/rds/PresignRequestHandlerTest.java [hunk 3]
```diff
@@ -141,7 +141,7 @@ public void testSourceRegionRemovedFromOriginalRequest() throws URISyntaxExcepti
     private SdkHttpFullRequest marshallRequest(CopyDBSnapshotRequest request) throws URISyntaxException {
         SdkHttpFullRequest marshalled = SdkHttpFullRequestAdapter.toHttpFullRequest(marshaller.marshall(request));
         return marshalled.toBuilder()
-                         .endpoint(new DefaultServiceEndpointBuilder(RDSClient.SERVICE_NAME, Protocol.HTTPS.toString())
+                         .endpoint(new DefaultServiceEndpointBuilder("rds", Protocol.HTTPS.toString())
```
Comment (shorea) [inline services/rds/src/test/java/software/amazon/awssdk/services/rds/PresignRequestHandlerTest.java]: Why the change here?

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java (hunk 1 above)
Comment (spfink) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java]: The type is not obvious if we support Apache Async.

You can't easily get the client name from the dependencies as ApplyUserAgentStage doesn't know if the sync or async client is being used though presumably only 1 will be non-null.

File: http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpClient.java (hunk 2 above)
Comment (spfink) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpClient.java]: I can add documentation but their will always be the possibility of skewed metrics. I think having it potentially inaccurate is still better than not having the data. Hopefully people will use good values if well documented.

Will add sanitization.

File: services/rds/src/test/java/software/amazon/awssdk/services/rds/PresignRequestHandlerTest.java (hunk 3 above)
Comment (spfink) [inline services/rds/src/test/java/software/amazon/awssdk/services/rds/PresignRequestHandlerTest.java]: SERVICE_NAME was renamed to SIGNING_NAME.

I could re-add a concept of SERVICE_NAME but not sure where it would be used.

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java (hunk 1 above)
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java]: > The type is not obvious if we support Apache Async.

CLIENT_NAME = "ApacheAsync"

> You can't easily get the client name from the dependencies as ApplyUserAgentStage doesn't know if the sync or async client is being used though presumably only 1 will be non-null.

Hmm I see, I was thinking about making that an either then it should be easy to pull out the right one.

Comment (spfink) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java]: Having a client name of ApacheAsync would work but I switched to separate values as then we could at least guarantee that Sync vs Async would be correct as we hardcode it into AmazonHttpClient and AmazonAsyncHttpClient. That way even if the client name metric is not always perfect because it is up to the implementor, sync vs async will always be correct.

File: services/rds/src/test/java/software/amazon/awssdk/services/rds/PresignRequestHandlerTest.java (hunk 3 above)
Comment (shorea) [inline services/rds/src/test/java/software/amazon/awssdk/services/rds/PresignRequestHandlerTest.java]: Hmm should this be using endpoint prefix? For rds and most services it's the same but for this usage it seems to make more sense.

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java (hunk 1 above)
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java]: I don't think it's worth the complexity. We will choose good names for our clients and it's up to others to use good names for theirs. Another idea is have the implementation append it that way they have control whether is reported on or not (to guard against the potentially crappy client name)

File: services/rds/src/test/java/software/amazon/awssdk/services/rds/PresignRequestHandlerTest.java (hunk 3 above)
Comment (spfink) [inline services/rds/src/test/java/software/amazon/awssdk/services/rds/PresignRequestHandlerTest.java]: Moved to using ServiceMetadata

**Comment by spfink** [comment]
Closing this out. Still want to do it but is very out of date and will just be easier to redo it from scratch.
